File: services/idun_agent_standalone_ui/lib/api/traces.ts
/**
 * Typed wrappers for /admin/api/v1/traces.
 *
 * Hand-rolled types mirroring libs/idun_agent_schema/src/idun_agent_schema/
 * standalone/traces.py (the StandaloneTrace and StandaloneSpan models).
 * The backend uses `to_camel` alias generator with `populate_by_name=True`,
 * so the wire format is camelCase — these TS interfaces match the JSON
 * keys exactly.
 *
 * All fetchers go through the shared `apiFetch` from `./client`, the same
 * wrapper the rest of the admin pages use (cookie credentials, SSO
 * bearer header, 401 redirect, ApiError envelope).
 */

import { apiFetch } from "./client";

// ── Types ───────────────────────────────────────────────────────────────

/**
 * One row in `GET /admin/api/v1/traces`.
 *
 * `otelTraceId` is the lowercase hex encoding of the 16-byte W3C trace
 * id; the DB stores raw bytes and the router converts at the boundary.
 */
export type StandaloneTraceListItem = {
  otelTraceId: string;
  name: string;
  startedAt: string;
  endedAt: string | null;
  latencyMs: number | null;
  totalTokens: number | null;
  totalCostUsd: number | null;
  models: string[];
  status: string | null;
  userId: string | null;
  sessionId: string | null;
  tags: string[];
};

/**
 * One span row from `standalone_span`.
 *
 * `otelSpanId` and `parentSpanId` are 8-byte hex strings. `otelTraceId`
 * is the trailing 8-byte slice of the parent trace id (the span table
 * stores half — see the exporter).
 */
export type StandaloneSpanRead = {
  otelSpanId: string;
  otelTraceId: string;
  parentSpanId: string | null;
  name: string;
  kind: string;
  startedAt: string;
  endedAt: string | null;
  latencyMs: number | null;
  model: string | null;
  provider: string | null;
  promptTokens: number | null;
  completionTokens: number | null;
  cacheReadTokens: number | null;
  cacheWriteTokens: number | null;
  totalTokens: number | null;
  costUsd: number | null;
  costBreakdown: Record<string, unknown> | null;
  costSource: string | null;
  status: string | null;
  attributes: Record<string, unknown> | null;
  events: Array<Record<string, unknown>> | null;
};

/**
 * Recursive node carrying a span and its children. Depth capped at 32
 * by the router (see schema `§28`).
 */
export type StandaloneSpanTreeNode = {
  span: StandaloneSpanRead;
  children: StandaloneSpanTreeNode[];
};

/** Body of `GET /admin/api/v1/traces/{otelTraceId}`. */
export type StandaloneTraceDetail = {
  trace: StandaloneTraceListItem;
  tree: StandaloneSpanTreeNode[];
};

/**
 * Query filters for `GET /admin/api/v1/traces`.
 *
 * Pagination is cursor-based on `(startedAt DESC, otelTraceId)`; the
 * `cursor` is an opaque base64url payload the router round-trips.
 */
export type StandaloneTraceListFilters = {
  startedAfter?: string;
  startedBefore?: string;
  model?: string;
  status?: string;
  userId?: string;
  sessionId?: string;
  nameContains?: string;
  limit?: number;
  cursor?: string;
};

/** Filter shape for bulk delete — mirrors list filters minus pagination. */
export type StandaloneTraceBulkDeleteFilters = Omit<
  StandaloneTraceListFilters,
  "limit" | "cursor"
>;

/** List response envelope. `nextCursor` is `null` on the last page. */
export type StandaloneTraceListResponse = {
  items: StandaloneTraceListItem[];
  nextCursor: string | null;
  totalEstimate: number | null;
};

/**
 * Body of `GET /admin/api/v1/traces/_health`.
 *
 * When the trace pipeline is not attached the router returns zeroes
 * + `writerRunning: false` (a 404-free idle state for the UI panel).
 *
 * `databaseDialect` is the SQLAlchemy bind dialect name read off the
 * standalone DB engine (`"sqlite"`, `"postgresql"`, ...). It falls
 * back to `"unknown"` when the engine isn't reachable so the SQLite
 * banner only renders on a confirmed signal.
 */
export type StandaloneTraceHealth = {
  queueDepth: number;
  maxQueueSize: number;
  overflowCount: number;
  writerRunning: boolean;
  databaseDialect: string;
};

/**
 * Body of `DELETE /admin/api/v1/traces/{id}`.
 *
 * The router returns the `deleted: true` discriminator + the cascaded
 * span count so the UI can show "removed N spans" without a follow-up.
 */
export type StandaloneTraceDeleteResult = {
  deleted: true;
  deletedSpans: number;
};

/** Body of `DELETE /admin/api/v1/traces` (bulk by filter). */
export type StandaloneTraceBulkDeleteResult = {
  deletedTraces: number;
  deletedSpans: number;
};

// ── Fetchers ────────────────────────────────────────────────────────────

const BASE = "/admin/api/v1/traces";

/**
 * Shared TanStack Query cache key for `getTraceHealth`. Both the
 * `SqliteBanner` and the `PipelineHealthPanel` mount their own
 * `useQuery` with this key so the two consumers de-duplicate to a
 * single in-flight request and a single cache entry — addresses the
 * "two polls per 5s for the same endpoint" finding (#39).
 */
export const TRACE_HEALTH_QUERY_KEY = ["traces", "health"] as const;

/**
 * Strip undefined / null / empty-string entries from a record so a
 * downstream `URLSearchParams` doesn't leak `?model=&status=` to the
 * API. Centralised so future filter fields can't reintroduce the leak.
 */
export function pickDefined<T extends Record<string, unknown>>(
  params: T,
): Partial<T> {
  const out: Partial<T> = {};
  for (const [key, value] of Object.entries(params) as [keyof T, unknown][]) {
    if (value === undefined || value === null || value === "") continue;
    out[key] = value as T[keyof T];
  }
  return out;
}

function buildQuery(params: Record<string, string | number | undefined>): string {
  const filtered = pickDefined(params);
  const usp = new URLSearchParams();
  for (const [key, value] of Object.entries(filtered)) {
    usp.set(key, String(value));
  }
  const qs = usp.toString();
  return qs ? `?${qs}` : "";
}

export function listTraces(
  filters: StandaloneTraceListFilters = {},
): Promise<StandaloneTraceListResponse> {
  return apiFetch<StandaloneTraceListResponse>(`${BASE}${buildQuery(filters)}`);
}

export function getTrace(otelTraceId: string): Promise<StandaloneTraceDetail> {
  return apiFetch<StandaloneTraceDetail>(
    `${BASE}/${encodeURIComponent(otelTraceId)}`,
  );
}

export function deleteTrace(
  otelTraceId: string,
): Promise<StandaloneTraceDeleteResult> {
  return apiFetch<StandaloneTraceDeleteResult>(
    `${BASE}/${encodeURIComponent(otelTraceId)}`,
    { method: "DELETE" },
  );
}

export function bulkDeleteTraces(
  filters: StandaloneTraceBulkDeleteFilters = {},
): Promise<StandaloneTraceBulkDeleteResult> {
  return apiFetch<StandaloneTraceBulkDeleteResult>(
    `${BASE}${buildQuery(filters)}`,
    { method: "DELETE" },
  );
}

export function getTraceHealth(): Promise<StandaloneTraceHealth> {
  return apiFetch<StandaloneTraceHealth>(`${BASE}/_health`);
}

export const tracesApi = {
  listTraces,
  getTrace,
  deleteTrace,
  bulkDeleteTraces,
  getTraceHealth,
};

File: libs/idun_agent_schema/src/idun_agent_schema/standalone/traces.py
"""Standalone trace + span admin contracts.

The trace admin REST surface is collection-scoped and addresses
individual traces by their hex-encoded W3C ``otel_trace_id``. The
storage layer keeps the id as raw bytes (16 bytes for the trace row,
an 8-byte slice for span rows); the wire format is hex everywhere so
JSON stays clean.

Schema source: ``tasks/trace-feature-08-05-2026/README.md`` decision
``§29`` (admin REST shape) and ``§28`` (depth-limited recursive CTE
for span tree assembly).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from ._base import _CamelModel


class StandaloneTraceListItem(_CamelModel):
    """One row in ``GET /admin/api/v1/traces`` list response.

    ``otel_trace_id`` is the lowercase hex encoding of the 16-byte
    W3C trace id. The DB stores raw bytes; routers convert at the
    boundary so JSON consumers never see a binary blob.
    """

    otel_trace_id: str
    name: str
    started_at: datetime
    ended_at: datetime | None = None
    latency_ms: float | None = None
    total_tokens: int | None = None
    total_cost_usd: float | None = None
    models: list[str] = Field(default_factory=list)
    status: str | None = None
    user_id: str | None = None
    session_id: str | None = None
    tags: list[str] = Field(default_factory=list)


class StandaloneSpanRead(_CamelModel):
    """One span read from the standalone_span table.

    ``otel_span_id`` and ``parent_span_id`` are 8-byte hex strings.
    ``otel_trace_id`` is the 8-byte slice of the parent trace id (the
    span table only stores the trailing half — see the exporter).
    """

    otel_span_id: str
    otel_trace_id: str
    parent_span_id: str | None = None
    name: str
    kind: str
    started_at: datetime
    ended_at: datetime | None = None
    latency_ms: float | None = None
    model: str | None = None
    provider: str | None = None
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    cache_read_tokens: int | None = None
    cache_write_tokens: int | None = None
    total_tokens: int | None = None
    cost_usd: float | None = None
    cost_breakdown: dict[str, Any] | None = None
    cost_source: str | None = None
    status: str | None = None
    attributes: dict[str, Any] | None = None
    events: list[dict[str, Any]] | None = None


class StandaloneSpanTreeNode(_CamelModel):
    """Recursive node carrying a span and its children.

    The router builds the tree in Python from the flat recursive-CTE
    result. Depth is capped at 32 (decision ``§28``).
    """

    span: StandaloneSpanRead
    children: list[StandaloneSpanTreeNode] = Field(default_factory=list)


# Pydantic 2 needs an explicit rebuild so the forward reference resolves.
StandaloneSpanTreeNode.model_rebuild()


class StandaloneTraceDetail(_CamelModel):
    """Body of ``GET /admin/api/v1/traces/{otel_trace_id}``.

    ``tree`` holds the top-level spans of the trace. Orphan spans
    (those whose ``parent_span_id`` is not present in the same trace,
    e.g. truncated traces or cross-trace links) are also surfaced as
    roots so the UI never silently hides them.
    """

    trace: StandaloneTraceListItem
    tree: list[StandaloneSpanTreeNode] = Field(default_factory=list)


class StandaloneTraceListFilters(_CamelModel):
    """Query-parameter model for ``GET /admin/api/v1/traces``.

    Pagination is cursor-based on ``(started_at DESC, otel_trace_id)``;
    the ``cursor`` field is an opaque base64url-encoded payload the
    router round-trips without exposing the shape on the wire.
    """

    started_after: datetime | None = None
    started_before: datetime | None = None
    model: str | None = None
    status: str | None = None
    user_id: str | None = None
    session_id: str | None = None
    name_contains: str | None = None
    limit: int = Field(default=50, ge=1, le=500)
    cursor: str | None = None


class StandaloneTraceHealth(_CamelModel):
    """Body of ``GET /admin/api/v1/traces/_health``.

    Reflects the running ``StandaloneSpanExporter``'s queue. When the
    trace pipeline is not attached (engine booted without traces) the
    router returns zeroes + ``writer_running=False`` instead of 404 so
    the UI panel can render a stable "pipeline idle" state.

    ``database_dialect`` lets the UI conditionally render the SQLite
    operational banner without a second round-trip; it mirrors the
    SQLAlchemy bind dialect name (``"sqlite"`` or ``"postgresql"``)
    and falls back to ``"unknown"`` when no bind is reachable.
    """

    queue_depth: int
    max_queue_size: int
    overflow_count: int
    writer_running: bool
    database_dialect: str = "unknown"


class StandaloneTraceDeleteResult(_CamelModel):
    """Body of a single ``DELETE /admin/api/v1/traces/{id}`` response.

    Reports the deleted trace plus the cascaded span count so the UI
    can show "removed N spans" without a follow-up read.
    """

    deleted: Literal[True] = True
    deleted_spans: int


class StandaloneTraceBulkDeleteResult(_CamelModel):
    """Body of ``DELETE /admin/api/v1/traces`` (bulk by filter).

    The filter shape mirrors :class:`StandaloneTraceListFilters` (minus
    ``limit`` / ``cursor``); the response counts what was removed.
    """

    deleted_traces: int
    deleted_spans: int


class StandaloneTraceListResponse(_CamelModel):
    """Envelope for the list view.

    ``next_cursor`` is the opaque cursor for the next page or ``None``
    when the current page is the last one. ``total_estimate`` is left
    ``None`` on SQLite (count is expensive) and reserved for a fast
    PG estimate via ``pg_class.reltuples`` in a later iteration.
    """

    items: list[StandaloneTraceListItem] = Field(default_factory=list)
    next_cursor: str | None = None
    total_estimate: int | None = None
